Mathlib/Algebra/MonoidAlgebra/Defs.lean

```diff
@@ -87,183 +93,56 @@ def unexpander : Lean.PrettyPrinter.Unexpander
-    p f := Finsupp.induction_linear f zero add single
-
-end
-
-section Mul
-
-variable [Semiring k] [Mul G]
-
-/-- The multiplication in a monoid algebra. We make it irreducible so that Lean doesn't unfold
-it trying to unify two things that are different. -/
-@[irreducible] def mul' (f g : MonoidAlgebra k G) : MonoidAlgebra k G :=
-  f.sum fun a₁ b₁ => g.sum fun a₂ b₂ => single (a₁ * a₂) (b₁ * b₂)
-
-/-- The product of `f g : MonoidAlgebra k G` is the finitely supported function
-whose value at `a` is the sum of `f x * g y` over all pairs `x, y`
-such that `x * y = a`. (Think of the group ring of a group.) -/
-instance instMul : Mul (MonoidAlgebra k G) := ⟨MonoidAlgebra.mul'⟩
-
-/-- The product of `f g : k[G]` is the finitely supported function
-whose value at `a` is the sum of `f x * g y` over all pairs `x, y`
-such that `x + y = a`. (Think of the product of multivariate
-polynomials where `α` is the additive monoid of monomial exponents.) -/
-instance _root_.AddMonoidAlgebra.instMul [Add G] : Mul (AddMonoidAlgebra k G) :=
-  ⟨fun f g => MonoidAlgebra.mul' (G := Multiplicative G) f g⟩
-
-attribute [to_additive existing instMul] instMul
-
-@[to_additive (dont_translate := k) mul_def]
-theorem mul_def {f g : MonoidAlgebra k G} :
-    f * g = f.sum fun a₁ b₁ => g.sum fun a₂ b₂ => single (a₁ * a₂) (b₁ * b₂) := by
-  with_unfolding_all rfl
+lemma ofCoeff_inj {x y : M →₀ R} : ofCoeff x = ofCoeff y ↔ x = y := coeffEquiv.symm.injective.eq_iff
 
-@[to_additive (dont_translate := k)]
-instance nonUnitalNonAssocSemiring : NonUnitalNonAssocSemiring (MonoidAlgebra k G) :=
-  { Finsupp.instAddCommMonoid with
-    -- Porting note: `refine` & `exact` are required because `simp` behaves differently.
-    left_distrib := fun f g h => by
-      haveI := Classical.decEq G
-      simp only [mul_def]
-      refine Eq.trans (congr_arg (sum f) (funext₂ fun a₁ b₁ => sum_add_index ?_ ?_)) ?_ <;>
-        simp only [mul_add, mul_zero, single_zero, single_add, forall_true_iff, sum_add]
-    right_distrib := fun f g h => by
-      haveI := Classical.decEq G
-      simp only [mul_def]
-      refine Eq.trans (sum_add_index ?_ ?_) ?_ <;>
-        simp only [add_mul, zero_mul, single_zero, single_add, forall_true_iff, sum_zero, sum_add]
-    zero_mul := fun f => by
-      simp only [mul_def]
-      exact sum_zero_index
-    mul_zero := fun f => by
-      simp only [mul_def]
-      exact Eq.trans (congr_arg (sum f) (funext₂ fun a₁ b₁ => sum_zero_index)) sum_zero }
+@[to_additive (attr := ext)] alias ⟨ext, _⟩ := coeff_inj
 
-end Mul
-
-section Semigroup
+@[to_additive] instance inhabited : Inhabited R[M] := coeffEquiv.inhabited
+@[to_additive] instance nontrivial [Nontrivial R] [Nonempty M] : Nontrivial R[M] :=
+  coeffEquiv.nontrivial
+@[to_additive] instance unique [Subsingleton R] : Unique R[M] := coeffEquiv.unique
+@[to_additive] instance addCommMonoid : AddCommMonoid R[M] :=
+  coeffEquiv.addCommMonoid
```

Comment:
```suggestion
  fast_instance% coeffEquiv.addCommMonoid
```
and for any other data-carrying instances you pull across

Reply:
Are you sure your advice is fully accurate? If I use it on `inhabited` or `unique`, it complains with
```
Provided instance does not reduce to a constructor application
  coeffEquiv.unique
Reduces to an application of Equiv.unique.

Use `set_option trace.Elab.fast_instance true` to analyze the error.

Trace of fields visited: []Lean 4
```

Reply:
`Equiv.unique` should be reducible, but isn't.